spec/abci/apps.md

```diff
@@ -451,3 +457,146 @@ If `appBlockHeight == storeBlockHeight`
     update the state using the saved ABCI responses but dont run the block against the real app.
 This happens if we crashed after the app finished Commit but before Tendermint saved the state.
 
+## State Sync
+
+A new node joining the network can simply join consensus at the genesis height and replay all
+historical blocks until it is caught up. However, for large chains this can take a significant
+amount of time, often on the order of weeks or months.
+
+State sync is an alternative mechanism for bootstrapping a new node, where it fetches a snapshot
+of the state machine at a given height and restores it. Depending on the application, this can
+be several orders of magnitude faster than replaying blocks.
+
+Note that state sync does not currently backfill historical blocks, so the node will have a 
+truncated block history - users are advised to consider the broader network implications of this in 
+terms of block availability and auditability. This functionality may be added in the future.
+
+For details on the specific ABCI calls and types, see the [methods and types section](abci.md).
+
+### Taking Snapshots
+
+Applications that want to support state syncing must take state snapshots at regular intervals. How
+this is accomplished is entirely up to the application. A snapshot consists of some metadata and
+a set of binary chunks in an arbitrary format:
+
+* `Height (uint64)`: The height at which the snapshot is taken. It must be taken after the given 
```

Comment:
Why before the next height is processed?

Reply:
Hmm, I still don't quite follow this. Before height 101 is processed by the state-machine? When would height H ever contain data from H+1? How would that even happen?

Reply:
It depends on how the application implements versioning, if at all. Consider a simple counter application that does something like this:

```go
counter += 1
go saveToFile(counter)
counter += 1
```

`saveToFile` may be called after the second `counter += 1`, in which case the snapshot would contain data from the next height. In this example the application should clearly not use a goroutine at all, but it's really up to the application how to implement this.

Reply:
I see. Well yes, that makes sense and I guess it should be noted. However, I don't see how or why this would ever happen in a deterministic state-machine (i.e. blockchain). But it is an important note nevertheless!

I would note though, that we rephrase this a bit to be a bit more of like your response. i.e. instead of saying before the next height, state that the snapshot height cannot data from the next height.

Reply:
We explicitly recommend users to take snapshots asynchronously, which complicates this. If the snapshot was taken synchronously (without a goroutine) then you're right, this would not happen - but then the chain would halt until the snapshot was completed, which can take a long time.

Reply:
Made some tweaks, let me know if anything is still unclear

https://github.com/tendermint/spec/pull/90/commits/2f4fc76477951a9a53fbf64d3a3c08ac02c2cbdc